### PR TITLE
Major Enhancements to User Authentication, Profile Management, and Search Functionality

### PR DESCRIPTION
**Pull request by C0nsumption**
This PR includes a series of enhancements and new features including an overhaul of the user authentication system, extended user model, enhanced signup process, implementation of logout system, user search functionality, and user interface enhancements. These changes improve the overall user experience and functionality of the platform.

### CODE REVIEW

File: server/main.py [hunk 1]
```diff
@@ -75,6 +101,11 @@ def decorated_function(*args, **kwargs):
 
     return decorated_function
 
+def allowed_file(filename):
```
Comment (PropeReferio) [inline server/main.py:104]: I would maybe write this like this:

def is_allowed_file(filename: str) -> bool:

What happens if the file is named `ashley.wilson.jpg`? Can you write a test for this edge case?
